server/mux.go

```diff
@@ -0,0 +1,259 @@
+	methods := allowedMethods[:]
+	if method != "*" {
+		um := strings.ToUpper(method)
+		var allowed bool
+		for _, am := range allowedMethods {
+			if um == am {
+				allowed = true
+				break
+			}
+		}
+		if !allowed {
+			panic(fmt.Errorf("method not allowed: %q, path: %q", um, path))
+		}
+
+		methods = []string{um}
+	}
+	return m.mustAddRoute(m.endpointRoot, methods, path, handler)
+}
+
+func (m *Mux) mustAddRoute(root *pathpattern.Node, methods []string, path string, handler http.Handler) *Mux {
+	const wildcardReplacement = "/{_couper_wildcardMatch*}"
+	const wildcardSearch = "/**"
+
+	// TODO: Unique Options per method if configurable later on
+	pathOptions := &pathpattern.Options{}
+
+	for _, method := range methods {
+		if strings.HasSuffix(path, wildcardSearch) {
+			pathOptions.SupportWildcard = true
+			path = path[:len(path)-len(wildcardSearch)] + wildcardReplacement
+		}
+
+		node, err := root.CreateNode(method+" "+path, pathOptions)
+		if err != nil {
+			panic(fmt.Errorf("create path node failed: %s %q: %v", method, path, err))
+		}
+
+		node.Value = &openapi3filter.Route{
+			Method:  method,
+			Path:    path,
+			Handler: handler,
+		}
+	}
+	return m
+}
+
+func (m *Mux) FindHandler(req *http.Request) http.Handler {
+	var route *openapi3filter.Route
+
+	node, paramValues := m.match(m.endpointRoot, req)
+	if node == nil {
+		// No matches for api or free endpoints. Determine if we have entered an api basePath
+		// and handle api related errors accordingly.
+		// Otherwise look for existing files or spa fallback.
+		if isConfigured(m.opts.APIBasePath) && m.isAPIError(req.URL.Path) {
+			return m.opts.APIErrTpl.ServeError(errors.APIRouteNotFound)
+		}
+
+		fileHandler, exist := m.hasFileResponse(req)
+		if exist {
+			return fileHandler
+		}
+
+		node, paramValues = m.match(m.spaRoot, req)
+		if node == nil {
+			if isConfigured(m.opts.FileBasePath) && m.isFileError(req.URL.Path) {
+				return m.opts.FileErrTpl.ServeError(errors.FilesRouteNotFound)
+			}
+
+			return m.opts.ServerErrTpl.ServeError(errors.Configuration)
+		}
+	}
+
+	route, _ = node.Value.(*openapi3filter.Route)
+
+	pathParams := make(request.PathParameter, len(paramValues))
+	paramKeys := node.VariableNames
+	for i, value := range paramValues {
+		key := paramKeys[i]
+		if strings.HasSuffix(key, "*") {
+			key = key[:len(key)-1]
+		}
+		pathParams[key] = value
+	}
+
+	ctx := req.Context()
+
+	const wcm = "_couper_wildcardMatch"
+	if wildcardMatch, ok := pathParams[wcm]; ok {
+		ctx = context.WithValue(ctx, request.Wildcard, wildcardMatch)
+		delete(pathParams, wcm)
+	}
+
+	ctx = context.WithValue(ctx, request.PathParams, pathParams)
+	*req = *req.Clone(ctx)
+
+	return route.Handler
+}
+
+func (m *Mux) match(root *pathpattern.Node, req *http.Request) (*pathpattern.Node, []string) {
+	matchPath := req.Method + " " + req.URL.Path
+	// no hosts are configured, lookup /wo hostPath first
+	node, paramValues := root.Match(matchPath)
```

Comment:
Hier muss die Reihenfolge geändert werden. Pattern mit einem Host sind präziser und sollten als erste in Betracht gezogen werden.